File: packages/python/cnos/src/cnos/graph.py
"""RuntimeGraph wire type and parse_runtime_graph() — mirrors Go's graph.go."""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

from cnos.errors import CnosError
from cnos.projection import OverrideSpec
from cnos.types import ConfigOrigin

GRAPH_ENV_VAR = "__CNOS_GRAPH__"


@dataclass
class GraphConfigEntry:
    key: str = ""
    value: Any = None
    namespace: str = ""
    source_id: str = ""
    plugin_id: str = ""
    workspace_id: str = ""
    profile: str = ""
    origin: Optional[ConfigOrigin] = None
    metadata: Dict[str, Any] = field(default_factory=dict)

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "GraphConfigEntry":
        return cls(
            key=d.get("key", ""),
            value=d.get("value"),
            namespace=d.get("namespace", ""),
            source_id=d.get("sourceId", ""),
            plugin_id=d.get("pluginId", ""),
            workspace_id=d.get("workspaceId", ""),
            profile=d.get("profile", ""),
            origin=ConfigOrigin.from_dict(d.get("origin")),
            metadata=dict(d.get("metadata") or {}),
        )


@dataclass
class GraphResolvedEntry:
    key: str = ""
    value: Any = None
    namespace: str = ""
    winner: GraphConfigEntry = field(default_factory=GraphConfigEntry)
    overridden: List[GraphConfigEntry] = field(default_factory=list)

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "GraphResolvedEntry":
        return cls(
            key=d.get("key", ""),
            value=d.get("value"),
            namespace=d.get("namespace", ""),
            winner=GraphConfigEntry.from_dict(d.get("winner") or {}),
            overridden=[GraphConfigEntry.from_dict(o) for o in (d.get("overridden") or [])],
        )


@dataclass
class GraphWorkspaceRoot:
    scope: str = ""
    workspace_id: str = ""
    path: str = ""

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "GraphWorkspaceRoot":
        return cls(
            scope=d.get("scope", ""),
            workspace_id=d.get("workspaceId", ""),
            path=d.get("path", ""),
        )


@dataclass
class GraphWorkspace:
    workspace_id: str = ""
    workspace_source: str = ""
    global_root: str = ""
    global_root_source: str = ""
    workspace_chain: List[str] = field(default_factory=list)
    workspace_roots: List[GraphWorkspaceRoot] = field(default_factory=list)

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "GraphWorkspace":
        return cls(
            workspace_id=d.get("workspaceId", ""),
            workspace_source=d.get("workspaceSource", ""),
            global_root=d.get("globalRoot", "") or "",
            global_root_source=d.get("globalRootSource", "") or "",
            workspace_chain=list(d.get("workspaceChain") or []),
            workspace_roots=[GraphWorkspaceRoot.from_dict(r) for r in (d.get("workspaceRoots") or [])],
        )


@dataclass
class RuntimeGraph:
    entries: List[GraphResolvedEntry] = field(default_factory=list)
    profile: str = ""
    resolved_at: str = ""
    profile_source: str = ""
    workspace: GraphWorkspace = field(default_factory=GraphWorkspace)
    overrides: Dict[str, OverrideSpec] = field(default_factory=dict)

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "RuntimeGraph":
        return cls(
            entries=[GraphResolvedEntry.from_dict(e) for e in (d.get("entries") or [])],
            profile=d.get("profile", ""),
            resolved_at=d.get("resolvedAt", ""),
            profile_source=d.get("profileSource", ""),
            workspace=GraphWorkspace.from_dict(d.get("workspace") or {}),
            overrides={k: OverrideSpec.from_dict(v) for k, v in (d.get("overrides") or {}).items()},
        )


def parse_runtime_graph(data: bytes) -> RuntimeGraph:
    """Parse and validate a runtime graph JSON payload."""
    try:
        raw = json.loads(data)
    except (json.JSONDecodeError, ValueError) as exc:
        raise CnosError(f"cnos: parse runtime graph: {exc}") from exc

    if not isinstance(raw, dict):
        raise CnosError("cnos: invalid runtime graph payload")

    graph = RuntimeGraph.from_dict(raw)

    if (
        not graph.profile
        or not graph.resolved_at
        or not graph.profile_source
        or not graph.workspace.workspace_id
        or not graph.workspace.workspace_source
        or graph.workspace.workspace_chain is None
        or graph.entries is None
    ):
        raise CnosError("cnos: invalid runtime graph payload")

    for entry in graph.entries:
        w = entry.winner
        if (
            not entry.key
            or not entry.namespace
            or not w.key
            or not w.namespace
            or not w.source_id
            or not w.plugin_id
            or not w.workspace_id
        ):
            raise CnosError("cnos: invalid runtime graph payload")

    return graph

File: packages/go/graph.go
package cnos

import (
	"encoding/json"
	"fmt"
	"sort"
)

const GraphEnvVar = "__CNOS_GRAPH__"

type RuntimeGraph struct {
	Entries       []GraphResolvedEntry    `json:"entries"`
	Profile       string                  `json:"profile"`
	ResolvedAt    string                  `json:"resolvedAt"`
	ProfileSource string                  `json:"profileSource"`
	Workspace     GraphWorkspace          `json:"workspace"`
	Overrides     map[string]OverrideSpec `json:"overrides,omitempty"`
}

type GraphResolvedEntry struct {
	Key        string             `json:"key"`
	Value      any                `json:"value"`
	Namespace  string             `json:"namespace"`
	Winner     GraphConfigEntry   `json:"winner"`
	Overridden []GraphConfigEntry `json:"overridden"`
}

type GraphConfigEntry struct {
	Key         string         `json:"key"`
	Value       any            `json:"value"`
	Namespace   string         `json:"namespace"`
	SourceID    string         `json:"sourceId"`
	PluginID    string         `json:"pluginId"`
	WorkspaceID string         `json:"workspaceId"`
	Profile     string         `json:"profile,omitempty"`
	Origin      *ConfigOrigin  `json:"origin,omitempty"`
	Metadata    map[string]any `json:"metadata,omitempty"`
}

type GraphWorkspace struct {
	WorkspaceID      string               `json:"workspaceId"`
	WorkspaceSource  string               `json:"workspaceSource"`
	GlobalRoot       string               `json:"globalRoot,omitempty"`
	GlobalRootSource string               `json:"globalRootSource,omitempty"`
	WorkspaceChain   []string             `json:"workspaceChain"`
	WorkspaceRoots   []GraphWorkspaceRoot `json:"workspaceRoots"`
}

type GraphWorkspaceRoot struct {
	Scope       string `json:"scope"`
	WorkspaceID string `json:"workspaceId"`
	Path        string `json:"path"`
}

func ParseRuntimeGraph(data []byte) (RuntimeGraph, error) {
	var graph RuntimeGraph
	if err := json.Unmarshal(data, &graph); err != nil {
		return RuntimeGraph{}, fmt.Errorf("cnos: parse runtime graph: %w", err)
	}

	if graph.Profile == "" || graph.ResolvedAt == "" || graph.ProfileSource == "" || graph.Workspace.WorkspaceID == "" || graph.Workspace.WorkspaceSource == "" || graph.Workspace.WorkspaceChain == nil || graph.Entries == nil {
		return RuntimeGraph{}, fmt.Errorf("cnos: invalid runtime graph payload")
	}

	for index := range graph.Entries {
		entry := &graph.Entries[index]
		if entry.Key == "" || entry.Namespace == "" || entry.Winner.Key == "" || entry.Winner.Namespace == "" || entry.Winner.SourceID == "" || entry.Winner.PluginID == "" || entry.Winner.WorkspaceID == "" {
			return RuntimeGraph{}, fmt.Errorf("cnos: invalid runtime graph payload")
		}
		if entry.Overridden == nil {
			entry.Overridden = []GraphConfigEntry{}
		}
	}

	if graph.Workspace.WorkspaceRoots == nil {
		graph.Workspace.WorkspaceRoots = []GraphWorkspaceRoot{}
	}

	return graph, nil
}

func runtimeEntryFromGraph(resolved GraphResolvedEntry) (*runtimeEntry, error) {
	entry := &runtimeEntry{
		key:       resolved.Key,
		namespace: resolved.Namespace,
		winner: runtimeProvenance{
			sourceID:    resolved.Winner.SourceID,
			pluginID:    resolved.Winner.PluginID,
			workspaceID: resolved.Winner.WorkspaceID,
			origin:      cloneOrigin(resolved.Winner.Origin),
		},
	}

	if promotedFrom, _ := resolved.Winner.Metadata["promotedFrom"].(string); promotedFrom != "" {
		entry.promotedFrom = promotedFrom
	}

	for _, override := range resolved.Overridden {
		entry.overridden = append(entry.overridden, runtimeProvenance{
			sourceID:    override.SourceID,
			pluginID:    override.PluginID,
			workspaceID: override.WorkspaceID,
			value:       override.Value,
			origin:      cloneOrigin(override.Origin),
		})
	}

	if resolved.Namespace == "secret" && isSecretReferenceValue(resolved.Value) {
		ref, err := toSecretReference(resolved.Value)
		if err != nil {
			return nil, err
		}
		if ref.Vault == "" {
			ref.Vault = "default"
		}
		entry.secretRef = &ref
		return entry, nil
	}

	if isDerivedValue(resolved.Value) {
		parsed, err := parseRawDerivedValue(resolved.Value)
		if err != nil {
			return nil, fmt.Errorf("cnos: parse derived formula for %s: %w", resolved.Key, err)
		}
		entry.formula = &parsed
		return entry, nil
	}

	entry.value = resolved.Value
	return entry, nil
}

func bootstrappedManifestFromGraph(graph RuntimeGraph) authoringManifest {
	namespaces := map[string]namespaceDefinition{}
	for key, value := range defaultNamespaceDefs {
		namespaces[key] = value
	}

	runtimeNamespaces := map[string]runtimeNamespaceDefinition{
		"process": {
			Description: "Live process runtime values.",
			ServerOnly:  true,
			BuiltIn:     true,
		},
	}
	for _, namespace := range discoverRuntimeNamespacesFromGraph(graph) {
		if namespace == "process" {
			continue
		}
		runtimeNamespaces[namespace] = runtimeNamespaceDefinition{ServerOnly: true}
	}

	frameworks := map[string]string{}
	for key, value := range defaultFrameworks {
		frameworks[key] = value
	}

	return authoringManifest{
		ProjectName:       "bootstrapped",
		WorkspaceDefault:  graph.Workspace.WorkspaceID,
		ProfileDefault:    graph.Profile,
		ResolveFrom:       []string{"default"},
		EnvMapping:        envMappingConfig{Explicit: map[string]string{}},
		Frameworks:        frameworks,
		Namespaces:        namespaces,
		RuntimeNamespaces: runtimeNamespaces,
		Vaults:            map[string]vaultDefinition{},
	}
}

func bootstrappedDynamicManifest() authoringManifest {
	namespaces := map[string]namespaceDefinition{}
	for key, value := range defaultNamespaceDefs {
		namespaces[key] = value
	}
	frameworks := map[string]string{}
	for key, value := range defaultFrameworks {
		frameworks[key] = value
	}
	return authoringManifest{
		ProjectName:      "dynamic",
		WorkspaceDefault: "base",
		EnvMapping:       envMappingConfig{Explicit: map[string]string{}},
		Frameworks:       frameworks,
		Namespaces:       namespaces,
		RuntimeNamespaces: map[string]runtimeNamespaceDefinition{
			"process": {
				Description: "Live process runtime values.",
				ServerOnly:  true,
				BuiltIn:     true,
			},
		},
		Vaults: map[string]vaultDefinition{},
		Schema: map[string]configSpecRule{},
	}
}

func discoverRuntimeNamespacesFromGraph(graph RuntimeGraph) []string {
	configNamespaces := map[string]struct{}{
		"value":  {},
		"secret": {},
		"meta":   {},
		"public": {},
	}
	for _, entry := range graph.Entries {
		configNamespaces[entry.Namespace] = struct{}{}
	}

	runtimeNamespaces := map[string]struct{}{}
	for _, entry := range graph.Entries {
		if !isDerivedValue(entry.Value) {
			continue
		}

		parsed, err := parseRawDerivedValue(entry.Value)
		if err != nil {
			continue
		}

		for _, ref := range parsed.refs {
			namespace := namespaceForKey(ref)
			if namespace == "" {
				continue
			}
			if _, ok := configNamespaces[namespace]; ok {
				continue
			}
			runtimeNamespaces[namespace] = struct{}{}
		}
	}

	namespaces := make([]string, 0, len(runtimeNamespaces))
	for namespace := range runtimeNamespaces {
		namespaces = append(namespaces, namespace)
	}
	sort.Strings(namespaces)
	return namespaces
}
